Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut

```diff
@@ -0,0 +1,269 @@
+global function AddDialogButton
+#elseif CLIENT
+global function ServerDialogueOnClientInit
+#elseif UI
+global function ClientCallback_RequestDialog
+#endif
+
+#if UI
+float cooldown = 0.0
+#endif
+
+#if SERVER
+
+struct  {
+    table< entity, array< void functionref( entity ) > > playerButtonCallBack // array for each button
+} file
+
+global struct ServerDialogButtonData
+{
+    void functionref( entity ) callback
+    string text
+}
+
+global struct ServerDialogData
+{
+    // var menu // cut no idea how this would work
+    string header = " "
+    string message = " "
+    // DialogMessageRuiData &ruiMessage
+    // array<int> messageColor = [161, 161, 161, 255] 
+    string image = "."
+    string rightImage = "."
+    bool forceChoice = false
+    bool noChoice = false
+    bool noChoiceWithNavigateBack = false
+    bool showSpinner = false
+    bool showPCBackButton = false
+    float inputDisableTime = 0
+    // table<int,bool> coloredButton
+    bool darkenBackground = false
+    bool useFullMessageHeight = false
+
+    array<ServerDialogButtonData> buttonData
+    // array<DialogFooterData> footerData // controller stuff?
+}
+
+void function ServerDialogInit()
+{
+    AddClientCommandCallback( "ns_button_dialog", ClientCommand_Button_Clicked )
+    AddCallback_OnClientConnected( _AddPlayerToCallbackTable )
+}
+
+void function _ResetPlayerButtonCallBacks( entity player )
+{
+    file.playerButtonCallBack[player].clear()
+}
+
+void function _AddPlayerButtonCallBack( entity player, void functionref( entity ) callback )
+{
+    file.playerButtonCallBack[player].append( callback )
+}
+
+void function _AddPlayerToCallbackTable( entity player )
+{
+    file.playerButtonCallBack[player] <- []
+}
+
+void function SendServerDialog( entity player, ServerDialogData dialog )
+{
+    array<string> args
+
+    args.append( StringReplace( dialog.header, " ", "/s", true, true ) )
+    args.append( StringReplace( dialog.message, " ", "/s", true, true ) )
+    args.append( dialog.image )
+    args.append( dialog.rightImage )
+    args.append( dialog.forceChoice.tointeger().tostring() )
+    args.append( dialog.noChoice.tointeger().tostring() )
+    args.append( dialog.noChoiceWithNavigateBack.tointeger().tostring() )
+    args.append( dialog.showSpinner.tointeger().tostring() )
+    args.append( dialog.showPCBackButton.tointeger().tostring() )
+    args.append( dialog.inputDisableTime.tostring() )
+    args.append( dialog.darkenBackground.tointeger().tostring() )
+    args.append( dialog.useFullMessageHeight.tointeger().tostring() )
+
+    _ResetPlayerButtonCallBacks( player )
+
+    for( int i; i < dialog.buttonData.len() && i < 4; i++ )
+    {
+        args.append( StringReplace( dialog.buttonData[i].text, " ", "/s", true, true ) )
+        _AddPlayerButtonCallBack( player, dialog.buttonData[i].callback )
+    }
+
+    while( args.len() < 16 ) // this wouldn't add buttons but I added it so the array len stays the same
+        args.append( "." )
+
```

Comment:
```
dialog.messageColor.resize( 4, 100 )
foreach(colour in dialog.messageColor )
    args.append( string(colour) )
 ```
 Since we cant use statically sized arrays, as respawn didnt implement them properly, we need to resize it like that